src/nfa/nfa.cc

```diff
@@ -517,46 +560,44 @@ void Mata::Nfa::remove_epsilon(Nfa* result, const Nfa& aut, Symbol epsilon)
 {
     assert(nullptr != result);
 
+    result->clear_nfa();
     result->increase_size(aut.get_num_of_states());
 
     // cannot use multimap, because it can contain multiple occurrences of (a -> a), (a -> a)
     StateMap<StateSet> eps_closure;
 
     // TODO: grossly inefficient
     // first we compute the epsilon closure
-    for (size_t i=0; i < aut.get_num_of_states(); ++i)
+    const size_t num_of_states{ aut.get_num_of_states() };
+    for (size_t i{ 0 }; i < num_of_states; ++i)
     {
         for (const auto& trans: aut[i])
         { // initialize
-            auto it_ins_pair = eps_closure.insert({i, {i}});
+            const auto it_ins_pair = eps_closure.insert({i, {i}});
             if (trans.symbol == epsilon)
             {
-                StateSet &closure = it_ins_pair.first->second;
+                StateSet& closure = it_ins_pair.first->second;
+                // TODO: Fix possibly insert to OrdVector. Create list already ordered, then merge (do not need to resize each time);
                 closure.insert(trans.states_to);
             }
         }
     }
 
     bool changed = true;
+    const TransSymbolStates epsilon_symbol_to_find{ epsilon };
     while (changed) { // compute the fixpoint
         changed = false;
-        for (size_t i=0; i < aut.get_num_of_states(); ++i)
-        {
-            for (auto const &trans: aut[i])
-            {
-                if (trans.symbol == epsilon)
-                {
-                    StateSet &src_eps_cl = eps_closure[i];
-                    for (State tgt : trans.states_to)
-                    {
-                        const StateSet &tgt_eps_cl = eps_closure[tgt];
-
-                        for (State st: tgt_eps_cl)
-                        {
-                            if (src_eps_cl.count(st) == 0) changed = true;
-                            src_eps_cl.insert(st);
-                        }
+        for (size_t i=0; i < num_of_states; ++i) {
+            const auto& state_transitions{ aut[i] };
+            const auto state_symbol_transitions{ state_transitions.find(epsilon_symbol_to_find) };
+            if (state_symbol_transitions != state_transitions.end()) {
+                StateSet &src_eps_cl = eps_closure[i];
+                for (const State tgt: state_symbol_transitions->states_to) {
+                    const StateSet &tgt_eps_cl = eps_closure[tgt];
+                    for (State st: tgt_eps_cl) {
+                        if (src_eps_cl.count(st) == 0) changed = true;
```

Comment:
Should there be `break` if the condition is true? This might save some computational time.

Reply:
I believe there could be. Nice catch.